### DISCUSSION TITLE
P/Invoke and UTF-32

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/runtime#98975] [Q&A] [answered]
On Linux, `wchar_t` is UTF-32. `CharSet.Unicode` is UTF-16.

**Comment by huoyaoyuan** [comment]
A UTF-32 char is just an `uint`. You can also use the more strongly-typed `System.Text.Rune`.

**Comment by colejohnson66** [comment]
What is your question? Is it how to marshal as UTF-32 in a `LibraryImport`?

**Reply by ghost** [reply]
Yes. My own research shows it's currently impossible. So, I have to come up with workarounds.

**Comment by tannergooding** [comment, marked as answer]
> On Linux, wchar_t is UTF-32.

On Windows `wchar_t` is notably UTF-16. It's one of the places where different platforms have different definitions for basic C/C++ types.

> CharSet.Unicode is UTF-16.

Yes, which matches the .NET `char` type. You would need to explicitly use `int` or `uint` (preferably matching the signedness of `wchar_t` for the underlying platform) when targeting Unix based platforms where `wchar_t` is 32-bits.
